File: docs/js/legend-panel.js
/**
 * Contributor legend panel.
 */

export class LegendPanel {
  constructor(element) {
    this.element = element;
  }

  update(authors, authorColors, totalCommits) {
    const sorted = Object.values(authors)
      .sort((a, b) => b.commitCount - a.commitCount);

    const rows = sorted.slice(0, 30).map(a => {
      const color = authorColors[a.email] || '#58a6ff';
      const pct = totalCommits > 0
        ? (a.commitCount / totalCommits * 100).toFixed(1) + '%'
        : '0%';

      return `<tr>
        <td><span class="legend-swatch" style="background: ${color}"></span></td>
        <td class="legend-name" style="color: ${color}">${escapeHtml(a.name)}</td>
        <td class="legend-count">${a.commitCount}</td>
        <td class="legend-pct">${pct}</td>
      </tr>`;
    }).join('');

    const more = sorted.length > 30
      ? `<tr><td colspan="4" style="color: var(--text-secondary)">... and ${sorted.length - 30} more</td></tr>`
      : '';

    this.element.innerHTML = `
      <h3 style="margin-bottom: 12px; color: var(--accent)">Contributors (${sorted.length})</h3>
      <table class="legend-table">
        <thead>
          <tr>
            <th></th>
            <th>Author</th>
            <th style="text-align: right">Commits</th>
            <th style="text-align: right">%</th>
          </tr>
        </thead>
        <tbody>${rows}${more}</tbody>
      </table>
      <div class="legend-summary">
        ${totalCommits} commits by ${sorted.length} authors
      </div>
    `;
  }
}

function escapeHtml(text) {
  const div = document.createElement('div');
  div.textContent = text;
  return div.innerHTML;
}

File: src/gitviz/tui/legend_panel.py
"""Contributor legend and statistics panel."""

from __future__ import annotations

from textual.widgets import Static
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from ..git.models import Author


class LegendPanel(Static):
    """Shows contributor legend with colors and stats."""

    def update_authors(
        self,
        authors: dict[str, Author],
        author_colors: dict[str, str],
        total_commits: int,
    ) -> None:
        """Refresh the legend with current author data."""
        sorted_authors = sorted(
            authors.values(), key=lambda a: a.commit_count, reverse=True
        )

        table = Table(
            title="Contributors",
            show_header=True,
            header_style="bold",
            expand=True,
            padding=(0, 1),
        )
        table.add_column("Color", width=3)
        table.add_column("Author", ratio=2)
        table.add_column("Commits", justify="right", width=8)
        table.add_column("%", justify="right", width=6)

        for author in sorted_authors[:25]:
            color = author_colors.get(author.email, "white")
            pct = (
                f"{author.commit_count / total_commits * 100:.1f}%"
                if total_commits
                else "0%"
            )
            table.add_row(
                Text("██", style=color),
                Text(author.name, style=color),
                str(author.commit_count),
                pct,
            )

        if len(sorted_authors) > 25:
            table.add_row(
                "", Text(f"... and {len(sorted_authors) - 25} more", style="dim"), "", ""
            )

        summary = Text()
        summary.append(f"\nTotal: {total_commits} commits by {len(sorted_authors)} authors\n", style="dim")

        panel = Panel(
            table,
            title=f"Contributors ({len(sorted_authors)})",
            border_style="magenta",
            subtitle="Press [L] to toggle",
        )
        self.update(panel)
